extensions/ql-vscode/src/qlpack-generator.ts

```diff
@@ -0,0 +1,85 @@
+import { mkdir, writeFile } from "fs-extra";
+import { dump } from "js-yaml";
+import { join } from "path";
+import { CodeQLCliServer } from "./cli";
+
+export type QueryLanguage =
+  | "csharp"
+  | "cpp"
+  | "go"
+  | "java"
+  | "javascript"
+  | "python"
+  | "ruby"
+  | "swift";
+
+export class QlPackGenerator {
+  private readonly qlpackName: string;
+  private readonly qlpackVersion: string;
+  private readonly header: string;
+  private readonly qlpackFileName: string;
+
+  constructor(
+    private readonly folderName: string,
+    private readonly queryLanguage: QueryLanguage,
+    private readonly cliServer: CodeQLCliServer,
+  ) {
+    this.qlpackName = `getting-started/codeql-extra-queries-${this.queryLanguage}`;
+    this.qlpackVersion = "1.0.0";
+    this.header = "# This is an automatically generated file.\n\n";
+
+    this.qlpackFileName = "qlpack.yml";
+  }
+
+  public async generate() {
+    await mkdir(this.folderName);
+
+    // create qlpack.yml
+    await this.createQlPackYaml();
+
+    // create example.ql
+    await this.createExampleQlFile();
+
+    // create codeql-pack.lock.yml
+    await this.createCodeqlPackLockYaml();
+  }
+
+  private async createQlPackYaml() {
+    const qlPackFile = join(this.folderName, this.qlpackFileName);
+
+    const qlPackYml = {
+      name: this.qlpackName,
+      version: this.qlpackVersion,
+      dependencies: {
+        [`codeql/${this.queryLanguage}-all`]: "*",
+      },
+    };
+
+    await writeFile(qlPackFile, this.header + dump(qlPackYml), "utf8");
+  }
+
+  private async createExampleQlFile() {
+    const exampleQlFile = join(this.folderName, "example.ql");
+
+    const exampleQl = `
```

Comment:
Does this query work for all languages? Maybe? Hopefully? I wonder if long term this is a template that should be provided by the language extractor since it could change in the future.

Reply:
Testing that these sample queries work for all languages will be difficult. It will be a poor user experience if the first time they try to generate a query, it has a compile error.

This is probably fine for now, but I don't know if this is a good long term solution.

Reply:
This query won't work for Ruby (where you need to either `import codeql.ruby.AST` or use `Ast::Block`). I would start by keeping the placeholder query much simpler, rather than trying to use the language libraries:
```
import <lang>
select "Hello, world!"
```

In future we can maintain example query files within the extension repo, and then we can have CI test that they still compile. Or even better, but more complicated: download one of the [example query packs](https://github.com/orgs/codeql/packages?q=example&tab=packages&q=examples), find a named example query within that, and copy its contents here.

Reply:
This is definitely safer. Maybe start with the simplest query and add a link in a comment to docs where you can learn about more sophisticated queries.

Future work to generate a more meaningful query.

Reply:
I had a look at what the language extractor is (Robert explained it to me) and I understand that would mean creating a new CLI command to tap into it and then use that command to generate our example file.

I'll make a note of this idea as I think it's a bit more involved to add a new command.

In the mean time I've gone the route of using Aditya's suggestion and simplify the example query. 

Thanks for the input! Really useful! Please keep it coming.